File: 01/node/solve.ts
import { readFileSync } from "fs";

const file = readFileSync(process.argv[2]).toString();

const lines = file.split("\n");
const delimeter = new RegExp(/\s+/);

const [leftArray, rightArray] = lines.reduce(
  ([resultLeft, resultRight], line) => {
    const [left, right] = line.split(delimeter);
    if (left && right) {
      resultLeft.push(+left);
      resultRight.push(+right);
    }
    return [resultLeft, resultRight];
  },
  [[], []] as [number[], number[]]
);

leftArray.sort();
rightArray.sort();

let sum = 0;

for (let i = 0; i < leftArray.length; i++) {
  const diff = Math.abs(leftArray[i] - rightArray[i]);
  sum += diff;
}

console.log("sum: ", sum);

File: 01/haskell/solve.hs
import System.Environment
import Data.List (sort)

diff :: Int -> Int -> Int
diff l r = abs (l - r)

sumDiffs :: [Int] -> [Int] -> Int
sumDiffs [] [] = 0
sumDiffs leftArray rightArray = sumDiffs (tail leftArray) (tail rightArray) + diff (head leftArray) (head rightArray) 

main = do
   args <- getArgs
   content <- readFile (head args)
   let linesOfTuples = map words (lines content)
   let leftArray = sort (map (!!0) linesOfTuples)
   let rightArray = sort (map (!!1) linesOfTuples)
   let sum = sumDiffs (map read leftArray ::[Int]) (map read rightArray ::[Int])
   print sum 
